micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmGcMetrics.java

```diff
@@ -242,8 +242,18 @@ private boolean nonGenerationalGcShouldUpdateDataSize(String gcName) {
     }
 
     private boolean isGenerationalGcConfigured() {
-        return ManagementFactory.getMemoryPoolMXBeans().stream().filter(JvmMemory::isHeap)
+        boolean result = false;
+
+        // Zing GPGC is always generational:
+        result |= ManagementFactory.getMemoryPoolMXBeans().stream().filter(JvmMemory::isHeap)
+                .map(MemoryPoolMXBean::getName).filter(name -> name.contains("GPGC")).findAny().isPresent();
+
+        // Keep original logic for other GCs:
+        // having more than 1 non-'tenured' pools is considered to be generational.
+        result |= ManagementFactory.getMemoryPoolMXBeans().stream().filter(JvmMemory::isHeap)
```

Comment:
Instead of doing these tricks with the `result` and going through the stream twice, would not it be better to write the filter in a way that can handle both?
If my understanding is correct, something like this?

```java
return ManagementFactory.getMemoryPoolMXBeans().stream()
                .filter(JvmMemory::isHeap)
                .map(MemoryPoolMXBean::getName)
                .filter(name -> name.contains("GPGC") || !name.contains("tenured"))
                .findAny()
                .isPresent();
```

Reply:
isPresent() is not enough, as the original code was looking for more than 1 !'tenured' beans.

So we can go with either something like this:

```
        // Zing GPGC is always generational and having more than 1 non-'tenured'
        // pools is also considered to be generational.
        int nonTenuredPools = 0;
        for (MemoryPoolMXBean bean : ManagementFactory.getMemoryPoolMXBeans()) {
            String name = bean.getName();
            if (!name.contains("tenured")) {
                nonTenuredPools++;
                if (nonTenuredPools == 2) {
                    return true;
                }
            }
            if (name.contains("GPGC")) {
                return true;
            }
        }
        return false;
```
or, slightly shorter, but little bit more cryptic and less efficient (still a single stream pass)
```
        // Zing GPGC is always generational and having more than 1 non-'tenured'
        // pools is also considered to be generational.
        return ManagementFactory.getMemoryPoolMXBeans().stream().filter(JvmMemory::isHeap)
                .map(MemoryPoolMXBean::getName)
                .mapToInt(name -> name.contains("GPGC") ? 2 : name.contains("tenured") ? 0 : 1)
                .sum() > 1;
```

Reply:
Yepp, you are right, it needs more than 1.
I like the more readable (/less cryptic) version with the for loop. Looking at all three options, I think I would go with that. What do you think?

Reply:
Agree with you, thanks! I'll update the diff